lib/qesapdeployment.pm

```diff
@@ -136,7 +144,104 @@ sub qesap_get_deployment_code {
         push(@log_files, QESAP_GIT_CLONE_LOG);
         assert_script_run("set -o pipefail ; $git_clone_cmd | tee " . QESAP_GIT_CLONE_LOG, quiet => 1);
     }
+    # Add symlinks for different provider directory naming between OpenQA and qesap-deployment
+    assert_script_run("ln -s " . TERRAFORM_DIR . "/aws " . TERRAFORM_DIR . "/ec2");
+    assert_script_run("ln -s " . TERRAFORM_DIR . "/gcp " . TERRAFORM_DIR . "/gce");
+}
+
+=head3 qesap_yaml_replace
+
+    Replaces yaml config file variables with parameters defined by OpenQA testode, yaml template or yaml schedule.
+    Openqa variables need to be added as a hash with key/value pair inside %run_args{openqa_variables}.
+    Example:
+        my %variables;
+        $variables{HANA_SAR} = get_required_var("HANA_SAR");
+        $variables{HANA_CLIENT_SAR} = get_required_var("HANA_CLIENT_SAR");
+        qesap_yaml_replace(openqa_variables=>\%variables);
+=cut
+
+sub qesap_yaml_replace {
+    my (%args) = @_;
+    my $variables = $args{openqa_variables};
+    my %variables = %$variables;
+    my %replaced_variables = ();
+
+    push(@log_files, QESAP_CONF_TRGT);
+
+    for my $variable (keys %variables) {
+        $replaced_variables{"%" . $variable . "%"} = %variables{$variable};
+    }
+    file_content_replace(QESAP_CONF_TRGT, %replaced_variables);
+    qesap_upload_logs();
+}
+
+
+=head3 qesap_execute
+
+    qesap_execute(cmd => $qesap_script_cmd [, verbose => 1] );
+
+    Execute qesap glue script commands. Check project documentation for available options:
+    https://github.com/SUSE/qe-sap-deployment
+=cut
+
+sub qesap_execute {
+    my (%args) = @_;
+    my $verbose = $args{verbose} ? "--verbose" : "";
+    my $qesap_cmd = DEPLOYMENT_DIR . "/scripts/qesap/qesap.py " . $verbose . " -c " . QESAP_CONF_TRGT . " -b " . DEPLOYMENT_DIR;
+    my $command = defined($args{cmd}) ? $args{cmd} : die("QESAP command to execute undefined");
```

Comment:
Suggestion:
```
die 'QESAP command to execute undefined' unless $args{cmd};
```
And then replace `$command` for `$args{cmd}` in lines 192, 195, 198 & 201.

BTW, `defined` checks for the value to be defined within the hash, so it returns true if said value is 0 or the empty string. As I don't suppose either of those can be a command, you can drop `defined`.

Reply:
OK thanks. Like previously I thought perl would not like it if for example $args{something} was not specified from calling function and I would point to this key.